homeassistant/components/bluetooth/storage.py

```diff
@@ -0,0 +1,223 @@
+                },
+                "service_data": {
+                    service_uuid: service_data.hex()
+                    for service_uuid, service_data in advertisement_data.service_data.items()
+                },
+                "service_uuids": advertisement_data.service_uuids,
+                "tx_power": advertisement_data.tx_power,
+                "rssi": advertisement_data.rssi,
+            },
+        }
+    return data
+
+
+def deserialize_discovered_device_advertisement_datas(
+    discovered_device_advertisement_datas: dict[str, dict[str, dict[str, Any]]]
+) -> dict[str, tuple[BLEDevice, AdvertisementData]]:
+    """Deserialize discovered_device_advertisement_datas."""
+    data: dict[str, tuple[BLEDevice, AdvertisementData]] = {}
+    for (
+        address,
+        device_advertisement_data,
+    ) in discovered_device_advertisement_datas.items():
+        device = device_advertisement_data["device"]
+        advertisement_data = device_advertisement_data["advertisement_data"]
+        data[address] = (
+            BLEDevice(  # type: ignore[no-untyped-call]
+                address=device["address"],
+                name=device["name"],
+                rssi=device["rssi"],
+                details=device["details"],
+            ),
+            AdvertisementData(
+                local_name=advertisement_data["local_name"],
+                manufacturer_data={
+                    manufacturer_id: bytes.fromhex(manufacturer_data)
+                    for manufacturer_id, manufacturer_data in advertisement_data[
+                        "manufacturer_data"
+                    ].items()
+                },
+                service_data={
+                    service_uuid: bytes.fromhex(service_data)
+                    for service_uuid, service_data in advertisement_data[
+                        "service_data"
+                    ].items()
+                },
+                service_uuids=advertisement_data["service_uuids"],
+                rssi=advertisement_data["rssi"],
+                tx_power=advertisement_data["tx_power"],
+                platform_data=(),
+            ),
+        )
+    return data
+
+
+def deserialize_discovered_device_timestamps(
+    discovered_device_timestamps: dict[str, float]
+) -> dict[str, float]:
+    """Deserialize discovered_device_timestamps."""
+    time_diff = time.time() - MONOTONIC_TIME()
+    return {
+        address: unix_time - time_diff
+        for address, unix_time in discovered_device_timestamps.items()
+    }
+
+
+def serialize_discovered_device_timestamps(
+    discovered_device_timestamps: dict[str, float]
+) -> dict[str, float]:
+    """Serialize discovered_device_timestamps."""
+    time_diff = time.time() - MONOTONIC_TIME()
+    return {
+        address: monotonic_time + time_diff
+        for address, monotonic_time in discovered_device_timestamps.items()
+    }
```

Comment:
We can likely move a lot of this code into `bluetooth_adapters` and add units for it to reduce the size of this PR